src/cascadia/WindowsTerminal/AppHost.cpp

```diff
@@ -215,49 +215,51 @@ void AppHost::_HandleSessionRestore(const bool startedForContent)
+        // a new window using the first saved layout information.
+        // The 2nd+ layout will always get a new window.
+        if (!_windowLogic.HasCommandlineArguments() &&
+            !_appLogic.HasSettingsStartupActions())
         {
-            uint32_t startIdx = 0;
-            // We want to create a window for every saved layout.
-            // If we are the only window, and no commandline arguments were provided
-            // then we should just use the current window to load the first layout.
-            // Otherwise create this window normally with its commandline, and create
-            // a new window using the first saved layout information.
-            // The 2nd+ layout will always get a new window.
-            if (!_windowLogic.HasCommandlineArguments() &&
-                !_appLogic.HasSettingsStartupActions())
-            {
-                _windowLogic.SetPersistedLayoutIdx(startIdx);
-                startIdx += 1;
-            }
+            _windowLogic.SetPersistedLayoutIdx(startIdx);
+            startIdx += 1;
+        }
 
-            // Create new windows for each of the other saved layouts.
-            for (const auto size = layouts.Size(); startIdx < size; startIdx += 1)
-            {
-                auto newWindowArgs = fmt::format(L"{0} -w new -s {1}", args.Commandline()[0], startIdx);
-
-                STARTUPINFO si;
-                memset(&si, 0, sizeof(si));
-                si.cb = sizeof(si);
-                wil::unique_process_information pi;
-
-                LOG_IF_WIN32_BOOL_FALSE(CreateProcessW(nullptr,
-                                                       newWindowArgs.data(),
-                                                       nullptr, // lpProcessAttributes
-                                                       nullptr, // lpThreadAttributes
-                                                       false, // bInheritHandles
-                                                       DETACHED_PROCESS | CREATE_UNICODE_ENVIRONMENT, // doCreationFlags
-                                                       nullptr, // lpEnvironment
-                                                       nullptr, // lpStartingDirectory
-                                                       &si, // lpStartupInfo
-                                                       &pi // lpProcessInformation
-                                                       ));
-            }
+        // Create new windows for each of the other saved layouts.
+        for (const auto size = layouts.Size(); startIdx < size; startIdx += 1)
+        {
+            auto newWindowArgs = fmt::format(L"{0} -w new -s {1}", args.Commandline()[0], startIdx);
+
+            STARTUPINFO si;
+            memset(&si, 0, sizeof(si));
+            si.cb = sizeof(si);
+            wil::unique_process_information pi;
+
+            LOG_IF_WIN32_BOOL_FALSE(CreateProcessW(nullptr,

```

Comment:
holy sh--, multi-window restore starts WindowsTerminal.exe _which then gets a handle to our Monarch and sends it a commandline_.

What the _heck_.

We could just... use Monarch directly here. I know you didn't write this code but I think we missed it when we migrated to single-proc.

Reply:
Yup. I felt like it would be too off-topic if I changed this part within the PR. But I agree that it's ripe for a refactoring.